File: bindings/junit/UnitTest.java
/**
 * Unit tests for Magnéto-Serge JUnit assertions (no server needed)
 */

import org.junit.jupiter.api.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.io.*;
import java.nio.file.*;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("Cassette Unit Tests")
public class UnitTest {

    private Path tempDir;

    @BeforeEach
    void setUp() throws IOException {
        tempDir = Files.createTempDirectory("magneto-test-");
    }

    @AfterEach
    void tearDown() throws IOException {
        // Cleanup temp directory
        Files.walk(tempDir)
            .sorted(Comparator.reverseOrder())
            .forEach(path -> {
                try {
                    Files.delete(path);
                } catch (IOException e) {
                    // Ignore
                }
            });
    }

    @Nested
    @DisplayName("Cassette Structure Tests")
    class CassetteStructureTests {

        @Test
        @DisplayName("Should create valid cassette JSON")
        void testValidCassetteStructure() throws IOException {
            String cassette = createTestCassette("test", 1);

            assertTrue(cassette.contains("\"version\""));
            assertTrue(cassette.contains("\"name\""));
            assertTrue(cassette.contains("\"interactions\""));
            assertTrue(cassette.contains("\"recorded_at\""));
        }

        @Test
        @DisplayName("Should have correct version")
        void testCassetteVersion() throws IOException {
            String cassette = createTestCassette("version-test", 1);

            assertTrue(cassette.contains("\"version\": \"1.0\""));
        }

        @Test
        @DisplayName("Should have correct name")
        void testCassetteName() throws IOException {
            String cassette = createTestCassette("name-test", 1);

            assertTrue(cassette.contains("\"name\": \"name-test\""));
        }
    }

    @Nested
    @DisplayName("HTTP Interaction Tests")
    class HttpInteractionTests {

        @Test
        @DisplayName("Should have HTTP interaction")
        void testHttpInteraction() throws IOException {
            String cassette = createTestCassette("http-test", 1);

            assertTrue(cassette.contains("\"Http\""));
            assertTrue(cassette.contains("\"request\""));
            assertTrue(cassette.contains("\"response\""));
        }

        @Test
        @DisplayName("Should have request method and URL")
        void testRequestStructure() throws IOException {
            String cassette = createTestCassette("request-test", 1);

            assertTrue(cassette.contains("\"method\": \"GET\""));
            assertTrue(cassette.contains("\"url\": \"https://api.example.com/users\""));
        }

        @Test
        @DisplayName("Should have response status and headers")
        void testResponseStructure() throws IOException {
            String cassette = createTestCassette("response-test", 1);

            assertTrue(cassette.contains("\"status\": 200"));
            assertTrue(cassette.contains("\"Content-Type\""));
            assertTrue(cassette.contains("\"application/json\""));
        }
    }

    @Nested
    @DisplayName("Cookie Tests")
    class CookieTests {

        @Test
        @DisplayName("Should have cookies array")
        void testHasCookies() throws IOException {
            String cassette = createTestCassette("cookie-test", 1);

            assertTrue(cassette.contains("\"cookies\""));
            assertTrue(cassette.contains("\"JSESSIONID\""));
        }

        @Test
        @DisplayName("Cookie should have required fields")
        void testCookieFields() throws IOException {
            String cassette = createTestCassette("cookie-fields-test", 1);

            // Required fields
            assertTrue(cassette.contains("\"name\": \"JSESSIONID\""));
            assertTrue(cassette.contains("\"value\": \"ABC123\""));
            assertTrue(cassette.contains("\"domain\": \"example.com\""));
            assertTrue(cassette.contains("\"path\": \"/\""));
            assertTrue(cassette.contains("\"secure\": true"));
            assertTrue(cassette.contains("\"http_only\": true"));
        }
    }

    @Nested
    @DisplayName("Multiple Interactions Tests")
    class MultipleInteractionsTests {

        @ParameterizedTest
        @ValueSource(ints = {1, 2, 5, 10})
        @DisplayName("Should handle varying interaction counts")
        void testMultipleInteractions(int count) throws IOException {
            String cassette = createTestCassette("multi-test", count);

            // Count occurrences of "Http" to verify interaction count
            int httpCount = countOccurrences(cassette, "\"Http\"");
            assertEquals(count, httpCount);
        }

        @Test
        @DisplayName("Should create multiple different cassettes")
        void testMultipleCassettes() throws IOException {
            String cassette1 = createTestCassette("cassette-1", 1);
            String cassette2 = createTestCassette("cassette-2", 2);

            assertTrue(cassette1.contains("\"name\": \"cassette-1\""));
            assertTrue(cassette2.contains("\"name\": \"cassette-2\""));

            int count1 = countOccurrences(cassette1, "\"Http\"");
            int count2 = countOccurrences(cassette2, "\"Http\"");

            assertEquals(1, count1);
            assertEquals(2, count2);
        }
    }

    @Nested
    @DisplayName("File Operations Tests")
    class FileOperationsTests {

        @Test
        @DisplayName("Should write cassette to file")
        void testWriteCassette() throws IOException {
            String cassette = createTestCassette("file-test", 1);
            Path filePath = tempDir.resolve("file-test.json");

            Files.writeString(filePath, cassette);

            assertTrue(Files.exists(filePath));
            assertTrue(Files.size(filePath) > 0);
        }

        @Test
        @DisplayName("Should read cassette from file")
        void testReadCassette() throws IOException {
            String originalCassette = createTestCassette("read-test", 2);
            Path filePath = tempDir.resolve("read-test.json");

            Files.writeString(filePath, originalCassette);
            String loadedCassette = Files.readString(filePath);

            assertEquals(originalCassette, loadedCassette);
        }

        @Test
        @DisplayName("Should handle multiple cassette files")
        void testMultipleCassetteFiles() throws IOException {
            for (int i = 0; i < 3; i++) {
                String cassette = createTestCassette("cassette-" + i, i + 1);
                Path filePath = tempDir.resolve("cassette-" + i + ".json");
                Files.writeString(filePath, cassette);
            }

            List<Path> files = Files.list(tempDir).toList();
            assertEquals(3, files.size());
        }
    }

    @Nested
    @DisplayName("JSON Validity Tests")
    class JsonValidityTests {

        @Test
        @DisplayName("Should be valid JSON")
        void testValidJson() {
            assertDoesNotThrow(() -> {
                String cassette = createTestCassette("json-test", 1);
                // If we can parse it successfully, it's valid JSON
                // (In real code, you'd use a JSON library here)
            });
        }

        @Test
        @DisplayName("Should have proper JSON structure")
        void testJsonStructure() throws IOException {
            String cassette = createTestCassette("structure-test", 1);

            assertTrue(cassette.startsWith("{"));
            assertTrue(cassette.endsWith("}\n"));
            assertTrue(cassette.contains("\"interactions\": ["));
            assertTrue(cassette.contains("\"cookies\": ["));
        }
    }

    // Helper methods

    private String createTestCassette(String name, int interactionCount) throws IOException {
        StringBuilder sb = new StringBuilder();

        sb.append("{\n");
        sb.append("  \"version\": \"1.0\",\n");
        sb.append("  \"name\": \"").append(name).append("\",\n");
        sb.append("  \"recorded_at\": \"2025-10-25T10:00:00Z\",\n");
        sb.append("  \"interactions\": [\n");

        for (int i = 0; i < interactionCount; i++) {
            if (i > 0) sb.append(",\n");
            sb.append("    {\n");
            sb.append("      \"kind\": {\n");
            sb.append("        \"Http\": {\n");
            sb.append("          \"request\": {\n");
            sb.append("            \"method\": \"GET\",\n");
            sb.append("            \"url\": \"https://api.example.com/users\",\n");
            sb.append("            \"headers\": {\"Accept\": \"application/json\"},\n");
            sb.append("            \"body\": null\n");
            sb.append("          },\n");
            sb.append("          \"response\": {\n");
            sb.append("            \"status\": 200,\n");
            sb.append("            \"headers\": {\"Content-Type\": \"application/json\"},\n");
            sb.append("            \"body\": [123, 34, 117, 115, 101, 114, 115, 34, 58, 91, 93, 125]\n");
            sb.append("          }\n");
            sb.append("        }\n");
            sb.append("      }\n");
            sb.append("    }");
        }

        sb.append("\n  ],\n");
        sb.append("  \"cookies\": [\n");
        sb.append("    {\n");
        sb.append("      \"name\": \"JSESSIONID\",\n");
        sb.append("      \"value\": \"ABC123\",\n");
        sb.append("      \"domain\": \"example.com\",\n");
        sb.append("      \"path\": \"/\",\n");
        sb.append("      \"expires\": null,\n");
        sb.append("      \"max_age\": null,\n");
        sb.append("      \"secure\": true,\n");
        sb.append("      \"http_only\": true,\n");
        sb.append("      \"same_site\": null,\n");
        sb.append("      \"created_at\": \"2025-10-25T10:00:00Z\"\n");
        sb.append("    }\n");
        sb.append("  ]\n");
        sb.append("}\n");

        return sb.toString();
    }

    private int countOccurrences(String text, String pattern) {
        int count = 0;
        int index = 0;

        while ((index = text.indexOf(pattern, index)) != -1) {
            count++;
            index += pattern.length();
        }

        return count;
    }
}

File: bindings/phpunit/UnitTest.php
<?php
/**
 * Unit tests for Magnéto-Serge PHPUnit assertions (no server needed)
 */

use PHPUnit\Framework\TestCase;

class UnitTest extends TestCase
{
    private $tempDir;

    protected function setUp(): void
    {
        $this->tempDir = sys_get_temp_dir() . '/magneto-test-' . uniqid();
        mkdir($this->tempDir, 0777, true);
    }

    protected function tearDown(): void
    {
        $this->deleteDirectory($this->tempDir);
    }

    // Cassette Structure Tests

    public function testValidCassetteStructure()
    {
        $cassette = $this->createTestCassette('test', 1);

        $this->assertStringContainsString('"version"', $cassette);
        $this->assertStringContainsString('"name"', $cassette);
        $this->assertStringContainsString('"interactions"', $cassette);
        $this->assertStringContainsString('"recorded_at"', $cassette);
    }

    public function testCassetteVersion()
    {
        $cassette = $this->createTestCassette('version-test', 1);

        $this->assertStringContainsString('"version": "1.0"', $cassette);
    }

    public function testCassetteName()
    {
        $cassette = $this->createTestCassette('name-test', 1);

        $this->assertStringContainsString('"name": "name-test"', $cassette);
    }

    // HTTP Interaction Tests

    public function testHttpInteraction()
    {
        $cassette = $this->createTestCassette('http-test', 1);

        $this->assertStringContainsString('"Http"', $cassette);
        $this->assertStringContainsString('"request"', $cassette);
        $this->assertStringContainsString('"response"', $cassette);
    }

    public function testRequestStructure()
    {
        $cassette = $this->createTestCassette('request-test', 1);

        $this->assertStringContainsString('"method": "GET"', $cassette);
        $this->assertStringContainsString('"url": "https://api.example.com/users"', $cassette);
    }

    public function testResponseStructure()
    {
        $cassette = $this->createTestCassette('response-test', 1);

        $this->assertStringContainsString('"status": 200', $cassette);
        $this->assertStringContainsString('"Content-Type"', $cassette);
        $this->assertStringContainsString('"application/json"', $cassette);
    }

    // Cookie Tests

    public function testHasCookies()
    {
        $cassette = $this->createTestCassette('cookie-test', 1);

        $this->assertStringContainsString('"cookies"', $cassette);
        $this->assertStringContainsString('"JSESSIONID"', $cassette);
    }

    public function testCookieFields()
    {
        $cassette = $this->createTestCassette('cookie-fields-test', 1);

        $this->assertStringContainsString('"name": "JSESSIONID"', $cassette);
        $this->assertStringContainsString('"value": "ABC123"', $cassette);
        $this->assertStringContainsString('"domain": "example.com"', $cassette);
        $this->assertStringContainsString('"path": "/"', $cassette);
        $this->assertStringContainsString('"secure": true', $cassette);
        $this->assertStringContainsString('"http_only": true', $cassette);
    }

    // Multiple Interactions Tests

    /**
     * @dataProvider interactionCountProvider
     */
    public function testMultipleInteractions($count)
    {
        $cassette = $this->createTestCassette('multi-test', $count);

        $httpCount = substr_count($cassette, '"Http"');
        $this->assertEquals($count, $httpCount);
    }

    public function interactionCountProvider()
    {
        return [
            [1],
            [2],
            [5],
            [10],
            [50]
        ];
    }

    public function testMultipleCassettes()
    {
        $cassette1 = $this->createTestCassette('cassette-1', 1);
        $cassette2 = $this->createTestCassette('cassette-2', 2);

        $this->assertStringContainsString('"name": "cassette-1"', $cassette1);
        $this->assertStringContainsString('"name": "cassette-2"', $cassette2);

        $count1 = substr_count($cassette1, '"Http"');
        $count2 = substr_count($cassette2, '"Http"');

        $this->assertEquals(1, $count1);
        $this->assertEquals(2, $count2);
    }

    // File Operations Tests

    public function testWriteCassette()
    {
        $cassette = $this->createTestCassette('file-test', 1);
        $filePath = $this->tempDir . '/file-test.json';

        file_put_contents($filePath, $cassette);

        $this->assertFileExists($filePath);
        $this->assertGreaterThan(0, filesize($filePath));
    }

    public function testReadCassette()
    {
        $originalCassette = $this->createTestCassette('read-test', 2);
        $filePath = $this->tempDir . '/read-test.json';

        file_put_contents($filePath, $originalCassette);
        $loadedCassette = file_get_contents($filePath);

        $this->assertEquals($originalCassette, $loadedCassette);
    }

    public function testMultipleCassetteFiles()
    {
        for ($i = 0; $i < 3; $i++) {
            $cassette = $this->createTestCassette('cassette-' . $i, $i + 1);
            $filePath = $this->tempDir . '/cassette-' . $i . '.json';
            file_put_contents($filePath, $cassette);
        }

        $files = glob($this->tempDir . '/*.json');
        $this->assertCount(3, $files);
    }

    // JSON Validity Tests

    public function testValidJson()
    {
        $cassette = $this->createTestCassette('json-test', 1);

        $decoded = json_decode($cassette, true);

        $this->assertNotNull($decoded);
        $this->assertIsArray($decoded);
        $this->assertEquals(JSON_ERROR_NONE, json_last_error());
    }

    public function testJsonStructure()
    {
        $cassette = $this->createTestCassette('structure-test', 1);
        $data = json_decode($cassette, true);

        $this->assertArrayHasKey('version', $data);
        $this->assertArrayHasKey('name', $data);
        $this->assertArrayHasKey('interactions', $data);
        $this->assertArrayHasKey('cookies', $data);
        $this->assertArrayHasKey('recorded_at', $data);
    }

    public function testInteractionsArray()
    {
        $cassette = $this->createTestCassette('array-test', 3);
        $data = json_decode($cassette, true);

        $this->assertIsArray($data['interactions']);
        $this->assertCount(3, $data['interactions']);
    }

    public function testCookiesArray()
    {
        $cassette = $this->createTestCassette('cookies-array-test', 1);
        $data = json_decode($cassette, true);

        $this->assertIsArray($data['cookies']);
        $this->assertGreaterThan(0, count($data['cookies']));
    }

    // Helper Methods

    private function createTestCassette($name, $interactionCount)
    {
        $interactions = [];

        for ($i = 0; $i < $interactionCount; $i++) {
            $interactions[] = [
                'kind' => [
                    'Http' => [
                        'request' => [
                            'method' => 'GET',
                            'url' => 'https://api.example.com/users',
                            'headers' => ['Accept' => 'application/json'],
                            'body' => null
                        ],
                        'response' => [
                            'status' => 200,
                            'headers' => ['Content-Type' => 'application/json'],
                            'body' => [123, 34, 117, 115, 101, 114, 115, 34, 58, 91, 93, 125]
                        ]
                    ]
                ]
            ];
        }

        $cassette = [
            'version' => '1.0',
            'name' => $name,
            'recorded_at' => '2025-10-25T10:00:00Z',
            'interactions' => $interactions,
            'cookies' => [
                [
                    'name' => 'JSESSIONID',
                    'value' => 'ABC123',
                    'domain' => 'example.com',
                    'path' => '/',
                    'expires' => null,
                    'max_age' => null,
                    'secure' => true,
                    'http_only' => true,
                    'same_site' => null,
                    'created_at' => '2025-10-25T10:00:00Z'
                ]
            ]
        ];

        return json_encode($cassette, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . "\n";
    }

    private function deleteDirectory($dir)
    {
        if (!file_exists($dir)) {
            return;
        }

        $files = array_diff(scandir($dir), ['.', '..']);

        foreach ($files as $file) {
            $path = $dir . '/' . $file;
            is_dir($path) ? $this->deleteDirectory($path) : unlink($path);
        }

        rmdir($dir);
    }
}
